File: library-infrastructure/src/main/java/com/dodecaedro/library/infrastructure/repository/JpaBorrowRepository.java
package com.dodecaedro.library.infrastructure.repository;

import com.dodecaedro.library.domain.model.Book;
import com.dodecaedro.library.domain.model.Borrow;
import com.dodecaedro.library.domain.model.User;
import com.dodecaedro.library.domain.repository.BorrowRepository;
import com.dodecaedro.library.infrastructure.projection.BorrowExcerptProjection;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import org.springframework.data.rest.core.annotation.RepositoryRestResource;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@RepositoryRestResource(
		collectionResourceRel = "borrows",
		path = "borrows",
		excerptProjection = BorrowExcerptProjection.class
)
public interface JpaBorrowRepository extends CrudRepository<Borrow, UUID>, BorrowRepository {

	@CacheEvict({"books", "users"})
	Borrow save(Borrow borrow);

	Optional<Borrow> findByBookAndUser(Book book, User user);

	@Query("""
			SELECT COUNT(b) FROM Borrow b JOIN b.user u
			WHERE u = :user
			AND b.actualReturnDate IS NULL
			AND :time > b.expectedReturnDate
			""")
	int countExpiredBorrows(@Param("user") User user, @Param("time") ZonedDateTime time);

	@Query("""
			FROM Borrow b
			JOIN FETCH b.user u
			JOIN FETCH b.book k
			WHERE b.actualReturnDate IS NULL
			AND :time > b.expectedReturnDate
			""")
	List<Borrow> findAllExpiredBorrows(@Param("time") ZonedDateTime time);

	@Query("""
			SELECT COUNT(b) FROM Borrow b JOIN b.user u
			WHERE u = :user
			AND b.actualReturnDate IS NULL
			""")
	int countActiveBorrows(User user);
}

File: library-infrastructure/src/main/kotlin/com/dodecaedro/library/infrastructure/repository/JpaBorrowRepository.kt
package com.dodecaedro.library.infrastructure.repository

import com.dodecaedro.library.domain.model.Book
import com.dodecaedro.library.domain.model.Borrow
import com.dodecaedro.library.domain.model.User
import com.dodecaedro.library.domain.repository.BorrowRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.PagingAndSortingRepository
import org.springframework.data.repository.query.Param
import org.springframework.data.rest.core.annotation.RepositoryRestResource
import java.time.ZonedDateTime
import java.util.*

@RepositoryRestResource(
	collectionResourceRel = "borrows",
	path = "borrows"
)
interface JpaBorrowRepository: PagingAndSortingRepository<Borrow, UUID>, BorrowRepository {
	@CacheEvict("books", "users")
	override fun save(borrow: Borrow): Borrow

	override fun findByBookAndUser(book: Book, user: User): Borrow?

	@Query("""
			SELECT COUNT(b) FROM Borrow b JOIN b.user u
			WHERE u = :user
			AND b.actualReturnDate IS NULL
			AND :time > b.expectedReturnDate
			""")
	override fun countExpiredBorrows(@Param("user") user: User, @Param("time") nowDate: ZonedDateTime): Int

	@Query("""
			FROM Borrow b
			JOIN FETCH b.user u
			JOIN FETCH b.book k
			WHERE b.actualReturnDate IS NULL
			AND :time > b.expectedReturnDate
			""")
	override fun findAllExpiredBorrows(@Param("time") time: ZonedDateTime): List<Borrow>

	@Query("""
			SELECT COUNT(b) FROM Borrow b JOIN b.user u
			WHERE u = :user
			AND b.actualReturnDate IS NULL
			""")
	override fun countActiveBorrows(@Param("user") user: User): Int
}
